driver/src/orderbook/streamed/state.rs

```diff
@@ -84,97 +85,204 @@ impl State {
+        sell_balance.sell(event.executed_sell_amount, block_batch_id)?;
+        buy_balance.buy(event.executed_buy_amount, block_batch_id)?;
+
+        *order = new_order;
+        self.balances.insert(sell_balance_key, sell_balance);
+        self.balances.insert(buy_balance_key, buy_balance);
+
+        Ok(())
+    }
+
+    fn apply_trade_reversion(
+        &mut self,
+        event: &TradeReversion,
+        block_batch_id: BatchId,
+    ) -> Result<(), Error> {
+        let order = self
+            .orders
+            .get_mut(&(event.owner, event.order_id))
+            .ok_or(Error::UnknownOrder(event.order_id))?;
+        let sell_token = self
+            .tokens
+            .get_address_by_id(order.sell_token)
+            .ok_or(Error::UnknownToken(order.sell_token))?;
+        let buy_token = self
+            .tokens
+            .get_address_by_id(order.buy_token)
+            .ok_or(Error::UnknownToken(order.buy_token))?;
+
+        let mut new_order = *order;
+        let sell_balance_key = (event.owner, sell_token);
+        let mut sell_balance = *self
+            .balances
+            .get(&sell_balance_key)
+            .ok_or(Error::RevertingNonExistentTrade)?;
+        let buy_balance_key = (event.owner, buy_token);
+        let mut buy_balance = *self
+            .balances
+            .get(&buy_balance_key)
+            .ok_or(Error::RevertingNonExistentTrade)?;
+
+        new_order.revert_trade(event.executed_sell_amount, block_batch_id)?;
+        sell_balance.revert_sell(event.executed_sell_amount, block_batch_id)?;
+        buy_balance.revert_buy(event.executed_buy_amount, block_batch_id)?;
+
+        *order = new_order;
+        self.balances.insert(sell_balance_key, sell_balance);
+        self.balances.insert(buy_balance_key, buy_balance);
+
+        Ok(())
+    }
+
+    fn apply_solution_submission(
+        &mut self,
+        event: &SolutionSubmission,
+        block_batch_id: BatchId,
+    ) -> Result<(), Error> {
+        let fee_token = self
+            .tokens
+            .get_address_by_id(0)
+            .ok_or(Error::SolutionWithoutFeeToken)?;
+        match self.last_solution.batch_id.cmp(&block_batch_id) {
+            Ordering::Less => {
+                let balance = self
+                    .balances
+                    .entry((self.last_solution.submitter, fee_token))
+                    .or_default();
+                balance.add_balance(self.last_solution.burnt_fees)?;
+            }
+            Ordering::Equal => (),
+            Ordering::Greater => return Err(Error::SolutionForPastBatch),
+        }
+
+        // Make sure that the submitter gets an entry an balances so their balance shows up when
+        // place an order for the fee token. Otherwise this would not happen because we only create
+        // balances on trades.
+        self.balances
+            .entry((event.submitter, fee_token))
+            .or_default();
```

Comment:
I don't understand this reasoning. What's the advantage of having a default entry in the balance list? Wouldn't we be able to lazily create it on read?